src/eliater/workflow.py

```diff
@@ -0,0 +1,122 @@
+from y0.algorithm.falsification import get_conditional_independencies
+from y0.dsl import Variable
+from y0.graph import NxMixedGraph, set_latent
+from y0.struct import get_conditional_independence_tests
+
+
+def get_state_space_map(
+    data: pd.DataFrame, threshold: Optional[int] = 10
+) -> Dict[Variable, Literal["discrete", "continuous"]]:
+    """Get a dictionary from each variable to its type."""
+    column_values_unique_count = {
+        column_name: data[column_name].nunique() for column_name in data.columns
+    }
+    return {
+        Variable(column): "discrete"
+        if column_values_unique_count[column] <= threshold
+        else "continuous"
+        for column in data.columns
+    }
+
+
+def is_data_discrete(data: pd.DataFrame) -> bool:
+    """Check if all the columns in the dataframe has discrete data."""
+    variable_types = set(get_state_space_map(data=data).values())
+    return variable_types == {"discrete"}
+
+
+def is_data_continuous(data: pd.DataFrame) -> bool:
+    """Check if all the columns in the dataframe has continuous data."""
+    variable_types = set(get_state_space_map(data).values())
+    return variable_types == {"continuous"}
+
+
+def choose_default_test(data: pd.DataFrame) -> str:
+    """Choose the default statistical test for testing conditional independencies based on the data."""
+    if is_data_discrete(data):
+        return "chi-square"
+    if is_data_continuous(data):
+        return "pearson"
+    raise NotImplementedError(
+        "Mixed data types are not allowed. Either all of the columns of data should be discrete / continuous."
+    )
+
+
+def fix_graph(
+    graph: NxMixedGraph,
+    data: pd.DataFrame,
+    test: Optional[str] = None,
+    significance_level: Optional[float] = 0.05,
+) -> NxMixedGraph:
+    """Repairs the graph by adding undirected edges between the nodes that fail the conditional independency test."""
+    if not test:
+        test = choose_default_test(data)
+
+    tests = get_conditional_independence_tests()
+    if test not in tests:
+        raise ValueError(f"`{test}` is invalid. Supported CI tests are: {sorted(tests)}")
+
+    if is_data_continuous(data) and test != "pearson":
+        warnings.warn(
+            message="The data is continuous. Either discretize and use chi-square or use the pearson.",
+            stacklevel=2,
+        )
+
+    if is_data_discrete(data) and test == "pearson":
+        raise Exception("Cannot run pearson on discrete data. Use chi-square instead.")
+
+    for conditional_independency in get_conditional_independencies(graph):
+        if not conditional_independency.test(
+            data, boolean=True, method=test, significance_level=significance_level
+        ):
+            graph.add_undirected_edge(conditional_independency.left, conditional_independency.right)
+
+    return graph
+
+
+def find_all_nodes_in_causal_paths(
+    graph: NxMixedGraph, treatments: Union[Variable, Set[Variable]], outcome: Variable
```

Comment:
you should start a new PR for new lines of work

Reply:
also this is a good reminder that having everything in a supersized python file isn't always the best